File: c-examples/nativefilechooser-simple-app.c
#include <stdio.h>	// printf
#include <stdlib.h>	// exit,malloc
#include <string.h>	// strerror
#include <errno.h>	// errno
#include <Fl_C.h>
#include <Fl_WindowC.h>
#include <Fl_Sys_Menu_BarC.h>
#include <Fl_Native_File_ChooserC.h>
#include <Fl_BoxC.h>
#include <Fl_AskC.h>

typedef struct instance_data {
  fl_Native_File_Chooser* fc;
} instance_data;

int exist(const char *filename) {
  FILE *fp = fopen(filename, "r");
  if (fp) { fclose(fp); return(1); }
  else    { return(0); }
}
// 'Open' the file
void open_file(const char *filename) {
  printf("Open '%s'\n", filename);
}
void save(const char *filename) {
  printf("Saving '%s'\n", filename);
  if ( !exist(filename) ) {
    FILE *fp = fopen(filename, "w");				// create file if it doesn't exist
    if ( fp ) {
      // A real app would do something useful here.
      fprintf(fp, "Hello world.\n");
      fclose(fp);
    } else {
      flc_message_with_args("Error: %s: %s", filename, strerror(errno),NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
    }
  } else {
    // A real app would do something useful here.
  }
}
void open_cb(fl_Widget w, void *v) {
  fl_Window app = (fl_Window)v;
  instance_data* d = (instance_data*)Fl_Window_other_data(app);
  fl_Native_File_Chooser fc = d->fc;
  Fl_Native_File_Chooser_set_title(fc,"Open");
  // Fl_Native_File_Chooser_set_type(fc,BROWSE_FILE);
  switch ( Fl_Native_File_Chooser_show(fc) ) {
  case -1: break;	// Error
  case  1: break; 	// Cancel
  default:		// Choice
    Fl_Native_File_Chooser_set_preset_file(fc,Fl_Native_File_Chooser_filename(fc));
    open_file(Fl_Native_File_Chooser_filename(fc));
    break;
  }
}
void saveas_cb(fl_Widget w, void *v) {
  fl_Window app = (fl_Window)v;
  instance_data* d = (instance_data*)Fl_Window_other_data(app);
  fl_Native_File_Chooser fc = d->fc;
  Fl_Native_File_Chooser_set_title(fc,"Save As");
  Fl_Native_File_Chooser_set_type(fc,BROWSE_SAVE_FILE);
  switch ( Fl_Native_File_Chooser_show(fc) ) {
  case -1: break;	// Error
  case  1: break; 	// Cancel
  default:		// Choice
    Fl_Native_File_Chooser_set_preset_file(fc,Fl_Native_File_Chooser_filename(fc));
    save(Fl_Native_File_Chooser_filename(fc));
    break;
  }
}
void save_cb(fl_Widget w, void *v) {
  fl_Window app = (fl_Window)v;
  instance_data* d = (instance_data*)Fl_Window_other_data(app);
  fl_Native_File_Chooser fc = d->fc;
  if ( strlen(Fl_Native_File_Chooser_filename(fc)) == 0 ) {
    saveas_cb(w,v);
  } else {
    save(Fl_Native_File_Chooser_filename(fc));
  }
}

void quit_cb(fl_Widget w, void *v) {
  exit(0);
}
const char* untitled_default() {
  static char *filename = 0;
  if ( !filename ) {
    const char *home =
      getenv("HOME") ? getenv("HOME") :		 // unix
      getenv("HOME_PATH") ? getenv("HOME_PATH") :	 // windows
      ".";						 // other
    filename = (char*)malloc(strlen(home)+20);
    sprintf(filename, "%s/untitled.txt", home);
  }
  return(filename);
}

void initializeWindow(fl_Window window){
  Fl_Window_begin(window);
  instance_data* i = (instance_data*)malloc(sizeof(instance_data));
  fl_Native_File_Chooser chooser = (fl_Native_File_Chooser)Fl_Native_File_Chooser_New();
  Fl_Native_File_Chooser_set_filter(chooser,"Text\t*.txt\n");
  Fl_Native_File_Chooser_set_preset_file(chooser,untitled_default());
  i->fc = chooser;
  Fl_Window_set_other_data(window,i);
  Fl_Window_begin(window);
  fl_Sys_Menu_Bar menu = (fl_Sys_Menu_Bar)Fl_Sys_Menu_Bar_New(0,0,400,25);
  Fl_Sys_Menu_Bar_add_with_user_data(menu, "&File/&Open",  FL_COMMAND+'o', open_cb, (void*)window);
  Fl_Sys_Menu_Bar_add_with_user_data(menu,"&File/&Save",  FL_COMMAND+'s', save_cb, (void*)window);
  Fl_Sys_Menu_Bar_add_with_user_data(menu,"&File/&Save As", 0,  saveas_cb, (void*)window);
  Fl_Sys_Menu_Bar_add(menu,"&File/&Quit",  FL_COMMAND+'q', quit_cb);
  fl_Box box = (fl_Box)Fl_Box_New(20,25+20,Fl_Window_w(window)-40,Fl_Window_h(window)-40-25);
  Fl_Box_set_color(box,45);
  Fl_Box_set_box(box,FL_FLAT_BOX);
  Fl_Box_set_align(box,FL_ALIGN_CENTER|FL_ALIGN_INSIDE|FL_ALIGN_WRAP);
  Fl_Box_set_label(box,
  		   "This demo shows an example of implementing "
  		   "common 'File' menu operations like:\n"
  		   "    File/Open, File/Save, File/Save As\n"
  		   "..using the Fl_Native_File_Chooser widget.\n\n"
  		   "Note 'Save' and 'Save As' really *does* create files! "
  		   "This is to show how behavior differs when "
  		   "files exist vs. do not.");
  Fl_Box_set_labelsize(box,12);
  Fl_Window_end(window);
}
int main(int argc, char** argv){
  Fl_scheme("gtk+");
  fl_Window app = (fl_Window)Fl_Window_New_WithLabel(400,200,"Native File Chooser Example");
  initializeWindow(app);
  Fl_Window_show(app);
  return Fl_run();
}

File: src/Examples/nativefilechooser-simple-app.hs
module Main where
import qualified Graphics.UI.FLTK.LowLevel.FL as FL
import Graphics.UI.FLTK.LowLevel.Fl_Types
import Graphics.UI.FLTK.LowLevel.Fl_Enumerations
import Graphics.UI.FLTK.LowLevel.FLTKHS
import System.Directory
import System.Exit

openFile :: FilePath -> IO ()
openFile fp = print $ "Open '" ++ fp ++ "''"

saveFile :: FilePath -> IO ()
saveFile fp = do
  print $ "Saving '" ++ fp ++ "'"
  exists' <- doesFileExist fp
  if not exists'
    then writeFile fp "Hello world.\n"
    else return ()

openCb :: Ref NativeFileChooser -> Ref MenuItem ->  IO ()
openCb fc _ = do
  setTitle fc "Open"
  res' <- showWidget fc
  case res' of
   NativeFileChooserPicked -> do
     f' <- getFilename fc
     case f' of
      (Just f'') -> do
        setPresetFile fc f''
        openFile f''
      _ -> return ()
   _ -> return ()

saveAsCb :: Ref NativeFileChooser -> Ref MenuItem ->  IO ()
saveAsCb fc _ = do
  setTitle fc "Open"
  res' <- showWidget fc
  case res' of
   NativeFileChooserPicked -> do
     f' <- getFilename fc
     case f' of
      (Just f'') -> do
        setPresetFile fc f''
        saveFile f''
      _ -> return ()
   _ -> return ()

saveCb :: Ref NativeFileChooser -> Ref MenuItem ->  IO ()
saveCb fc w' = do
  f' <- getFilename fc
  case f' of
   Nothing -> saveAsCb fc w'
   (Just f'') -> saveFile f''

quitCb :: Ref MenuItem -> IO ()
quitCb _ = exitSuccess

initializeWindow :: Ref Window -> IO ()
initializeWindow w' = do
  chooser <- nativeFileChooserNew Nothing
  setFilter chooser "Text\t*.txt\n"
  setPresetFile chooser "untitiled.txt"
  begin w'
  menu <- sysMenuBarNew (toRectangle (0,0,400,25)) Nothing
  _ <- add menu "&File/&Open" (Just (KeySequence (ShortcutKeySequence [kb_CommandState] (NormalKeyType 'o')))) (Just (openCb chooser)) (MenuItemFlags [])
  _ <- add menu "&File/&Save" (Just (KeySequence (ShortcutKeySequence [kb_CommandState] (NormalKeyType 's')))) (Just (saveCb chooser)) (MenuItemFlags [])
  _ <- add menu "&File/&Save As" Nothing (Just (saveAsCb chooser)) (MenuItemFlags [])
  _ <- add menu "&File/&Quit" (Just (KeySequence (ShortcutKeySequence [kb_CommandState] (NormalKeyType 'q')))) (Just quitCb) (MenuItemFlags [])
  w_w' <- getW w'
  w_h' <- getH w'
  box' <- boxNew (toRectangle (20,25+20,w_w'-40,w_h'-40-25)) Nothing Nothing
  setColor box' (Color 45)
  setBox box' FlatBox
  setAlign box' (Alignments [AlignTypeCenter, AlignTypeInside, AlignTypeWrap])
  setLabel box' $ "This demo shows an example of implementing " ++
                  "common 'File' menu operations like:\n" ++
                  "    File/Open, File/Save, File/Save As\n" ++
                  "..using the Fl_Native_File_Chooser widget.\n\n" ++
                  "Note 'Save' and 'Save As' really *does* create files! " ++
                  "This is to show how behavior differs when " ++
                  "files exist vs. do not.";
  setLabelsize box' (FontSize 12)
  end w'

main :: IO ()
main = do
  _ <- FL.setScheme "gtk+"
  app <- windowNew (toSize (400,200)) Nothing (Just "Native File Chooser Example")
  initializeWindow app
  showWidget app
  _ <- FL.run
  return ()
